File: app/src/main/java/de/xikolo/managers/jobs/GetCourseJob.java
package de.xikolo.managers.jobs;

import android.support.annotation.Nullable;
import android.util.Log;

import com.birbit.android.jobqueue.Job;
import com.birbit.android.jobqueue.Params;
import com.birbit.android.jobqueue.RetryConstraint;

import org.greenrobot.eventbus.EventBus;

import de.xikolo.GlobalApplication;
import de.xikolo.managers.UserManager;
import de.xikolo.models.Course;
import de.xikolo.network.ApiV2Request;
import de.xikolo.utils.Config;
import de.xikolo.utils.NetworkUtil;
import io.realm.Realm;
import retrofit2.Response;

public class GetCourseJob extends Job {

    public static final String TAG = GetCourseJob.class.getSimpleName();

    private String courseId;

    public GetCourseJob(String courseId) {
        super(new Params(Priority.MID));

        this.courseId = courseId;
    }

    @Override
    public void onAdded() {
        if (Config.DEBUG) Log.i(TAG, TAG + " added | course.id " + courseId);
    }

    @Override
    public void onRun() throws Throwable {
        if (NetworkUtil.isOnline(GlobalApplication.getInstance())) {

            final Response<Course.JsonModel> response;

            if (UserManager.isLoggedIn()) {
                response = ApiV2Request.service().getCourse(UserManager.getTokenHeader()).execute();
            } else {
                response = ApiV2Request.service().getCourse(courseId).execute();
            }

            if (response.isSuccessful()) {
                if (Config.DEBUG) Log.i(TAG, "Course received");

                EventBus.getDefault().postSticky(new GetCourseJobEvent(NetworkJobEvent.State.SUCCESS, courseId));

                Realm realm = Realm.getDefaultInstance();
                realm.executeTransaction(new Realm.Transaction() {
                    @Override
                    public void execute(Realm realm) {
                        realm.copyToRealmOrUpdate(response.body().convertToRealmObject());
                    }
                });
                realm.close();
            } else {
                if (Config.DEBUG) Log.e(TAG, "Error while fetching course");
                EventBus.getDefault().postSticky(new GetCourseJobEvent(NetworkJobEvent.State.ERROR, courseId));
            }
        } else {
            EventBus.getDefault().postSticky(new GetCourseJobEvent(NetworkJobEvent.State.NO_NETWORK, courseId));
        }
    }

    @Override
    protected void onCancel(int cancelReason, @Nullable Throwable throwable) {
        EventBus.getDefault().postSticky(new GetCourseJobEvent(NetworkJobEvent.State.CANCEL, courseId));
    }

    @Override
    protected RetryConstraint shouldReRunOnThrowable(Throwable throwable, int runCount, int maxRunCount) {
        return RetryConstraint.CANCEL;
    }

    public static class GetCourseJobEvent extends NetworkJobEvent {

        public GetCourseJobEvent(State state, String id) {
            super(state, id);
        }

    }

}

File: app/src/main/java/de/xikolo/network/jobs/GetCourseJob.kt
package de.xikolo.network.jobs

import android.util.Log
import de.xikolo.config.Config
import de.xikolo.managers.UserManager
import de.xikolo.models.Course
import de.xikolo.models.Enrollment
import de.xikolo.network.ApiService
import de.xikolo.network.jobs.base.NetworkJob
import de.xikolo.network.sync.Sync
import de.xikolo.network.jobs.base.NetworkStateLiveData
import ru.gildor.coroutines.retrofit.awaitResponse

class GetCourseJob(private val courseId: String, networkState: NetworkStateLiveData, userRequest: Boolean) : NetworkJob(networkState, userRequest) {

    companion object {
        val TAG: String = GetCourseJob::class.java.simpleName
    }

    override suspend fun onRun() {
        val response = if (UserManager.isAuthorized) {
            ApiService.getInstance().getCourseWithEnrollment(courseId).awaitResponse()
        } else {
            ApiService.getInstance().getCourse(courseId).awaitResponse()
        }

        if (response.isSuccessful) {
            if (Config.DEBUG) Log.i(TAG, "Course received")

            Sync.Data.with(Course::class.java, response.body())
                .saveOnly()
                .run()
            Sync.Included.with(Enrollment::class.java, response.body())
                .addFilter("courseId", courseId)
                .run()

            success()
        } else {
            if (Config.DEBUG) Log.e(TAG, "Error while fetching course")
            error()
        }
    }

}
